src/nccl_ofi_rdma.c

```diff
@@ -1450,70 +1515,98 @@ static inline int process_completions(struct fi_cq_tagged_entry *cq_entry,
 
-static const char *req_type_str(nccl_net_ofi_rdma_req_type_t type)
-{
-	switch(type) {
-	case NCCL_OFI_RDMA_SEND_CONN:
-		return "SEND_CONN";
-	case NCCL_OFI_RDMA_SEND_CONN_RESP:
-		return "SEND_CONN_RESP";
-	case NCCL_OFI_RDMA_RECV_CONN:
-		return "RECV_CONN";
-	case NCCL_OFI_RDMA_RECV_CONN_RESP:
-		return "RECV_CONN_RESP";
-	case NCCL_OFI_RDMA_SEND:
-		return "SEND";
-	case NCCL_OFI_RDMA_RECV:
-		return "RECV";
-	case NCCL_OFI_RDMA_SEND_CTRL:
-		return "SEND_CTRL";
-	case NCCL_OFI_RDMA_RECV_SEGMS:
-		return "RECV_SEGMS";
-	case NCCL_OFI_RDMA_BOUNCE:
-		return "BOUNCE";
-	case NCCL_OFI_RDMA_FLUSH:
-		return "FLUSH";
-	case NCCL_OFI_RDMA_EAGER_COPY:
-		return "EAGER_COPY";
+	rc = fi_cq_readerr(rail->cq, &err_entry, 0);
+	if (OFI_UNLIKELY(rc == -FI_EAGAIN)) {
+		/*
+		 * Error not available yet.
+		 * fi_cq_read will keep returning -FI_EAVAIL so just bail out and try again later.
+		 */
+		return ncclSuccess;
+	} else if (OFI_UNLIKELY(rc < 0)) {
+		NCCL_OFI_WARN("Unable to read from fi_cq_readerr. RC: %zd. Error: %s",
+			      rc, fi_strerror(-rc));
+		ret = ncclSystemError;
+		goto exit;
 	}
-	return "unknown";
-}
 
-/*
- * @brief	Print NCCL OFI request information
- */
-static const char *nccl_net_ofi_req_str(nccl_net_ofi_rdma_req_t *req)
-{
-	static char buf[256];
-	snprintf(buf, sizeof(buf), "{ dev: %d, size: %zu, state: %s, type: %s }",
-		 req->dev_id,
-		 req->size,
-		 req_state_str(req->state),
-		 req_type_str(req->type)
-		);
-	return buf;
-}
+	if (err_entry.flags & FI_REMOTE_WRITE) {
+		req = get_req_from_imm_data(ep, err_entry.data);
+		if (!req) {
+			NCCL_OFI_WARN("Unknown remote write error");
+			ret = ncclSystemError;
+			goto exit;
+		}
+	} else {
+		/* For all other operations, ctx should be a req */
+		if (!err_entry.op_context) {
+			NCCL_OFI_WARN("Operation with NULL context completed with error!");
+			ret = ncclSystemError;
+			goto exit;
+		}
+		req = err_entry.op_context;
+	}
 
-static int post_rdma_ctrl(nccl_net_ofi_rdma_req_t *req);
+	NCCL_OFI_WARN("Request %p completed with error. RC: %d. Error: %s. Completed length: %ld, Request: %s",
+		      req, err_entry.err,
+		      fi_cq_strerror(rail->cq, err_entry.prov_errno, err_entry.err_data, NULL, 0),
+		      (long)err_entry.len, nccl_net_ofi_req_str(req));
+	if (req->type == NCCL_OFI_RDMA_BOUNCE) {
+		/* A bounce buffer receive failed -- this is an internal error so bail out */
+		NCCL_OFI_WARN("Fatal: Bounce buffer recv completed with error");
+	} else {
+		/* Move user-facing request to error state */
+		set_request_state_to_error(req);
+	}
 
-static int post_flush_req(nccl_net_ofi_rdma_req_t *req);
+	/*
+	 * Return appropriate errors up the stack.
+	 */
+	switch (err_entry.err) {
+	case -FI_EINVAL:
+		ret = ncclInternalError;
+		break;
+	case -FI_EMSGSIZE:
+		/* Per ext-net docs, this aligns with ncclInvalidUsage, which is
+		 * also defined in NCCL source, but for some reason that error
+		 * type is not available in err.h that we pull from ext-net
```

Comment:
I think Liran added it in his PR.

Reply:
Just saw that. Brian pointed out the same reason why I just dint add it here, we want to keep the upstream version intact.